test/test_nn.py

```diff
@@ -7192,6 +7192,73 @@ def m(t):
                     m = nn.Upsample(scale_factor=scale_factor, **kwargs).to(device)
                     _test_interpolate_helper(_make_input(3), scale_factor, m)
 
+    def _test_linear(self, input, m, in_sizes, out_sizes, expected_output_sizes):
+        in_list = [in_sizes] if isinstance(in_sizes, int) else list(in_sizes)
+        out_list = [out_sizes] if isinstance(out_sizes, int) else list(out_sizes)
+        sizes = out_list + in_list
+
+        self.assertEqual(m.weight.size(), torch.Size(sizes))
+        self.assertEqual(m.bias.size(), torch.Size(out_list))
+
+        def prod(sizes):
+            p = 1
+            for size in sizes:
+                p *= size
+            return p
+
+        weight = m.weight.data.clone().reshape(prod(out_list), prod(in_list))
+        bias = m.bias.data.clone().reshape(prod(out_list))
+
+        if in_list:
+            input_sizes = list(input.size())[:-len(in_list)]
+        else:
+            input_sizes = list(input.size())
+        input_sizes.append(prod(in_list))
+
+        output = torch.matmul(input.view(input_sizes), weight.t()) + bias
+
+        output_sizes = list(output.size())[:-1]
+        output_sizes.extend(out_list)
+        output = output.view(output_sizes)
+
+        self.assertEqual(output, m(input))
+        self.assertEqual(torch.Size(expected_output_sizes), m(input).size())
```

Comment:
can this ever fail if the check above passes?  I guess if you pass in the wrong expected_output_sizes, but why do you need to pass that in anyway?

Reply:
The test basically implements `linear` again by directly calling `matmul`. If the test's implementation of `linear` is incorrect, it is possible that `output` matches `m(input)` but both are wrong.

Let me add some tests cases that sets linear weight and bias, and then manually craft expected output values.